src/library/scala/collection/mutable/OpenHashMap.scala

```diff
@@ -91,42 +104,43 @@ extends AbstractMap[Key, Value]
     table = new Array[Entry](newSize)
     mask = newSize - 1
     oldTable.foreach( entry =>
```

Comment:
So I was going to take a quick look this morning, turns out it took a bit more time to figure out what's going on :) Here's my example:

```scala
class C {
  val table: Array[String] = null
  def t(): Unit = {
    table.foreach(addEntry)
  }
  def addEntry(s: String): Unit = ()
}
```

Basically the optimizer is supposed to inline foreach (the implementation is [inherited from `IndexedSeqOptimized`](https://github.com/scala/scala/blob/93f209dd65d5c05fc2cb61916a850940499c9261/src/library/scala/collection/IndexedSeqOptimized.scala#L28)), and the closure should be eliminated. But looking at the bytecode I saw something completely different, basically this:

```
        Function1 foreach_f = s -> {
            this.C$$$anonfun$1(s);
            return BoxedUnit.UNIT;
        };
        new ArrayOps.ofRef(Predef..MODULE$._refArrayOps((Object[])this.table())).iterator().foreach(foreach_f);
```

What the..? Turns out the inliner picked the wrong implementation and inserted the overridden [code from `IterableLike`](https://github.com/scala/scala/blob/93f209dd65d5c05fc2cb61916a850940499c9261/src/library/scala/collection/IterableLike.scala#L70). Here's the ticket: https://github.com/scala/scala-dev/issues/140.

But basically the 2.12.0 optimizer should be able to handle `Array.foreach` well. `Array.map` is a different story and will require much more sophisticated heuristics and stack allocation. I'll continue working on those for later 2.12 minor releases, but they won't be in 2.12.0.